src/broker/__tests__/fetchPartitionAllocation.spec.js

```diff
@@ -0,0 +1,83 @@
+jest.mock('../../utils/shuffle', () => jest.fn())
+
+const Broker = require('../index')
+const mockShuffle = require('../../utils/shuffle')
+const shuffle = jest.requireActual('../../utils/shuffle')
+
+const { createConnection, newLogger } = require('testHelpers')
+
+const minBytes = 1
+const maxBytes = 10485760 // 10MB
+const maxBytesPerPartition = 1048576 // 1MB
+const maxWaitTime = 100
+
+describe('Broker > Fetch', () => {
+  let broker, connection, fetchRequestMock
+
+  const createFetchTopic = (name, numPartitions) => {
+    const partitions = Array(numPartitions)
+      .fill()
+      .map((_, i) => ({
+        partition: i,
+        fetchOffset: 0,
+        maxBytes: maxBytesPerPartition,
+      }))
+
+    return {
+      topic: name,
+      partitions,
+    }
+  }
+
+  beforeEach(async () => {
+    mockShuffle.mockImplementation(shuffle)
+
+    connection = createConnection()
+    connection.send = jest.fn()
+    broker = new Broker({
+      connection,
+      logger: newLogger(),
+    })
+
+    broker.isConnected = () => true
+    broker.lookupRequest = () => fetchRequestMock
+  })
+
+  /**
+   * When requesting data for topic-partitions with more data than `maxBytes`,
+   * the response will be allocated in the order
+   * the topic-partitions appear in the request. To avoid biasing consumption,
+   * the order of all topic-partitions is randomized on each request.
+   *
+   * @see https://cwiki.apache.org/confluence/display/KAFKA/KIP-74%3A+Add+Fetch+Response+Size+Limit+in+Bytes
+   */
+  test('it should randomize the order of the requested topic-partitions', async () => {
+    const topics = [createFetchTopic('topic-a', 3), createFetchTopic('topic-b', 2)]
+
+    mockShuffle.mockImplementationOnce(() => [
```

Comment:
I don't really love this test, since I had to module mock `shuffle` - i.e. reach into the internals of `broker.fetch` - but it's hard to assert on things that have randomness, and in this case I really do want a specific order to check that consecutive partitions for the same topic get merged.

Reply:
Totally fair I'd say. My best attempt would asserting that the order is different, but even that's not guaranteed: ending up with the same order is still a valid shuffle result. Mocking this order is the closest I imagine you could get to verifying the logic in `broker.fetch`.

Reply:
Thinking about this a tiny bit more in case you'd really didn't like the test, I guess you could spy on `shuffle` instead of mocking it, so you'd know the order of topics it returned when called by `broker.fetch`. That would verify the interface between those two to work correctly, but since the surface area of that is so small and straightforward, I'm pretty sure that I wouldn't bother.

Reply:
The problem with that is that I need shuffle to actually return a very specific result in order to be able to validate that consecutive partitions for the same topic are merged. Meaning if we have topic-partitions:

```
topic-a | 1
topic-a | 2
topic-b | 1
topic-b | 2
```

And after shuffling it looks like:

```
topic-a | 1
topic-b | 1
topic-b | 2
topic-a | 2
```

We want the resulting request to be:

```
topic-a | 1
topic-b | 1, 2
topic-a | 2
```